docs/manual/resource-leak-checker.tex

```diff
@@ -0,0 +1,369 @@
+to write qualifiers from either type hierarchy. The most common annotations from
+these checkers that you might need to write are:
+
+\begin{description}
+
+\item[\refqualclasswithparams{checker/mustcall/qual}{MustCall}{String[] value}]
+for example on an element with compile-time type \<Object> that might contain a \<Socket>.
+See Section~\ref{must-call-annotations}.
+
+\item[\refqualclasswithparams{checker/mustcall/qual}{InheritableMustCall}{String[] value}]
+on any classes defined in your program that have must-call obligations. See Section~\ref{must-call-on-class}.
+
+\item[\refqualclass{checker/calledmethods/qual}{EnsuresCalledMethods}] on a method in
+your code that fulfills a must-call obligation of one of its parameters or of a field.
+See Section~\ref{called-methods-ensurescalledmethods}.
+
+\end{description}
+
+The Resource Leak Checker supports annotations that express
+aliasing patterns related to resource leaks:
+
+\begin{description}
+
+\item[\refqualclass{checker/mustcall/qual}{Owning}]
+\item[\refqualclass{checker/mustcall/qual}{NotOwning}]
+  expresses ownership.  When two aliases exist to the same Java object,
+  \<@Owning> and \<@NotOwning> indicate which of the two is responsible for
+  fulfilling must-call obligations.  For details, see
+  Section~\ref{resource-leak-ownership}.
+
+\item[\refqualclass{checker/mustcall/qual}{MustCallAlias}]
+  represents a ``resource-aliasing'' relationship.  Resource aliases are
+  distinct Java objects that control the same resource(s):
+  fulfilling the must-call obligations of one object also
+  fulfills the obligations of the other object.  For details,
+  see Section~\ref{resource-leak-resource-alias}.
+
+\end{description}
+
+The Resource Leak Checker also supports an annotation to permit re-assigning
+fields or re-opening resources:
+
+\begin{description}
+
+\item[\refqualclasswithparams{checker/mustcall/qual}{CreatesMustCallFor}{String value}]
+  is a declaration annotation that indicates that after a call to a method
+  with this annotation none of the must-call obligations of the in-scope, owning expression
+  listed in \<value> have been met.
+  In other words, the annotated method ``resets'' the must-call obligations of the expression.
+  Multiple \<@CreatesMustCallFor>
+  annotations can be written on the same method.  For more details
+  on how to use this annotation to permit re-assignment of owning
+  fields or the re-opening of resources, see Section~\ref{resource-leak-createsmustcallfor}.
+
+\end{description}
+
+\sectionAndLabel{Example of how safe resource usage is verified}{resource-leak-example}
+
+Consider the following example of a safe use of a \<Socket>:
+\begin{verbatim}
+Socket s = null;
+try {
+  s = new Socket(myHost, myPort);
+} catch (Exception e) {
+  // do nothing
+} finally {
+  if (s != null) {
+    s.close();
+  }
+}
+\end{verbatim}
+
+On the first line, \<s> is assigned to \<null>. At this point, it has a type from
```

Comment:
Thanks for writing up this example.  I found the paragraphs of text hard to follow, and I think it's mostly because they are too much text and because they made me keep referring to the code.  I suggest that you integrate the types with the code.

One way would be to put a comment on each line indicating the type of `s` after that line has executed.  Or, you could put comments between lines to indicate the type at those points in execution.  Then, the text could be much shorter:  it would only need to be written where a special explanation/discussion is necessary, not for every line.  Another thing that might be helpful is to number the lines.

Reply:
I've rewritten it based on your suggestion of putting the types in comments in the example. Let me know what you think - it might be possible to cut the text further, but I'd like to leave at least a little bit of exposition to guid the reader through what's happening in the figure.